lm/srilm.py

```diff
@@ -94,11 +95,103 @@ def hash(cls, kwargs):
         return super().hash(kwargs)
 
 
+class DiscountNgramsJob(Job):
+    """
+    Create a file with the discounted ngrams with SRILM
+    """
+
+    def __init__(
+        self,
+        ngram_order: int,
+        counts: tk.Path,
+        count_exe: tk.Path,
+        *,
+        vocab: Optional[tk.Path] = None,
+        extra_discount_args: Optional[List[str]] = None,
+        use_modified_srilm: bool = False,
+        cpu_rqmt: int = 1,
+        mem_rqmt: int = 48,
+        time_rqmt: float = 24,
+        fs_rqmt: str = "100G",
+    ):
+        """
+        :param ngram_order: order of the ngram counts, typically 3 or 4.
+        :param counts: file with the ngram counts, see :class:`CountNgramsJob.out_counts`.
+        :param vocab: vocabulary file for the discounting.
+        :param count_exe: path to the binary.
+        :param extra_discount_args: additional arguments for the discounting step.
+        :param use_modified_srilm: Use the i6 modified SRILM version by Sundermeyer.
+                                   The SRILM binary ngram-count was modified.
+                                   This version is currently only available internally.
+        :param cpu_rqmt: CPU requirements.
+        :param mem_rqmt: memory requirements.
+        :param time_rqmt: time requirements.
+        :param fs_rqmt: file size requirements.
+        """
+        self.ngram_order = ngram_order
+        self.counts = counts
+        self.vocab = vocab
+        self.discount_args = extra_discount_args or []
+        self.use_modified_srilm = use_modified_srilm
+
+        self.count_exe = count_exe
+
+        self.out_discounts = self.output_path("discounts", cached=True)
+
+        self.rqmt_run = {
+            "cpu": cpu_rqmt,
+            "mem": mem_rqmt,
+            "time": time_rqmt,
+            "qsub_args": f"-l h_fsize={fs_rqmt}",
+        }
+
+    def tasks(self):
+        yield Task("create_files", mini_task=True)
+        yield Task("run", rqmt=self.rqmt_run)
+
+    def create_files(self):
+        """creates bash script to compute discounts"""
+        cmd = [
+            f"{self.count_exe.get_path()} \\\n",
+            f"  -order {self.ngram_order} \\\n",
+        ]
+        if self.vocab is not None:
+            cmd += [
+                f"  -vocab {self.vocab.get_cached_path()} \\\n",
+            ]
+        cmd += ["  -kn discounts\\\n"] if not self.use_modified_srilm else [f"  -multi-kn-file discounts\\\n"]
+        cmd += [
+            f"  -read {self.counts.get_cached_path()} \\\n",
+            f"  {' '.join(self.discount_args)} -memuse\n",
+        ]
+
+        create_executable("run.sh", cmd)
+
+    def run(self):
+        """executes the previously created bash script and relinks outputs from work folder to output folder"""
+        subprocess.check_call("./run.sh")
+        relink("discounts", self.out_discounts.get_path())
```

Comment:
Unrelated to this PR, but could maybe be fixed here aswell: relink checks existance of the destination by checking for os.path.exists(dst) which will return false if the file the link is pointing to does not exist. I think it's probably not intended behavior. One solution would be to check also for `os.path.islink(dst)`. This is just FYI, fixing that is not required for approval.

Reply:
I added a check in the relink func